### DISCUSSION TITLE
How to query users in Keycloak and exclude those with a specific attribute?

### DISCUSSION THREAD

**Discussion by gediminasre** [keycloak/keycloak#21366] [Q&A]
Hello,

I'm currently working with Keycloak and need some assistance with querying users based on attributes while excluding those with a specific attribute value. I have explored the available options, but I couldn't find a straightforward solution.

In my attempt to filter users in Keycloak by excluding the ones with the attribute `USER_TYPE` set to `USER` I experimented with variations of the following queries: `/users?q=NOT(USER_TYPE:USER)`, `/users?q=NOT USER_TYPE:USER`, and `/users?q=NOT(USER_TYPE=USER)`. However, none of these queries provided the desired result.

If anyone has any insights or examples on how to construct the query correctly, incorporating negation to filter users based on attributes, I would greatly appreciate your guidance.

**Comment by danielFesenmeyer** [comment]
Negation is currently not supported in attribute queries, just equality checks are possible.

**Comment by dteleguin** [comment]
It is important to distinguish user properties (that map to UserModel fields) and custom user attributes. As `USER_TYPE` doesn't map to any UserModel fields, I'd speculate you're trying to query by custom user attributes. I'm afraid this is not supported in Keycloak. Your best option is to deploy a custom REST endpoint that would implement such query. There is a [method](https://github.com/keycloak/keycloak/blob/main/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L758) in Keycloak (used in a different context) that you can use as a starting point. Please also mind that the custom REST endpoint will need to be properly secured in order to prevent unauthorized access.

**Reply by gediminasre** [reply]
Yes, you are right, I am trying to query custom user attributes. I needed to allow admin users to manage only simple users (not other admins), and first thought was to add default filters in my BE service and exclude admins from all users list, but I went to another direction and restricted users accessibility implementing admin-fine-grained-authz
